File: ex2/ios/Classes/GmsPlacesApiImpl.swift
import Flutter
import GooglePlaces

extension FlutterError: Error {}

class GmsPlacesApiImpl: GmsPlacesApi {
    static var token: GMSAutocompleteSessionToken?
    static var fieldDict: [PlaceFields: GMSPlaceField] = [
        .formattedAddress: .formattedAddress,
        .addressComponents: .addressComponents,
        .businessStatus: .businessStatus,
        .placeID: .placeID,
        .coordinate: .coordinate,
        .name: .name,
        .photos: .photos,
        .plusCode: .plusCode,
        .types: .types,
        .viewport: .viewport,
    ]
    static var businessStatusDict: [GMSPlacesBusinessStatus: PlacesBusinessStatus] = [
        .closedPermanently: .closedPermanently,
        .closedTemporarily: .closedTemporarily,
        .operational: .operational,
        .unknown: .unknown,
    ]

    func ensureInitialized() throws {
        // アプリケーションのAppDelegate内でこのコードを実行することは、
        // アプリケーションの起動時にセッショントークンを初期化する
        // 一般的な方法です。AppDelegateのapplication(_
        // :didFinishLaunchingWithOptions:)メソッド内でこのコードを配置することで、アプリケーションが起動するたびに新しいセッショントークンが生成され、GmsPlacesApiImplクラスで使用されるようになります
        GmsPlacesApiImpl.token = GMSAutocompleteSessionToken()
    }

    func autocomplete(fromQuery: String, filter: PlacesFilterType, completion: @escaping (Result<[Prediction], Error>) -> Void) {

        let autocompleteFilter = GMSAutocompleteFilter()
        switch filter {
        case .address:
            autocompleteFilter.type = .address
        case .city:
            autocompleteFilter.type = .city
        case .establishment:
            autocompleteFilter.type = .establishment
        case .geocode:
            autocompleteFilter.type = .geocode
        case .region:
            autocompleteFilter.type = .region
        }
        if GmsPlacesApiImpl.token == nil {
            let error = FlutterError(code: "emptyToken", message: "emptyToken", details: "emptyToken")
            print("1")
            completion(.failure(error))
        }
        GMSPlacesClient.shared().findAutocompletePredictions(
            fromQuery: fromQuery,
            filter: autocompleteFilter,
            sessionToken: GmsPlacesApiImpl.token
            
        ) { items, error in
            if error != nil {
                print("2")
                print(error as Any)
                completion(.failure(FlutterError(code: "searchError", message: error?.localizedDescription, details: "")))
            } else {
                if items == nil {
                    print("3")
                    completion(.success([]))
                    return
                }
                var result = [Prediction]()
                for item in items! {
                    result.append(Prediction(
                        attributed: PredictionAttributed(
                            fullText: item.attributedFullText.string,
                            primaryText: item.attributedPrimaryText.string,
                            secondaryText: item.attributedSecondaryText?.string
                        ),
                        placeID: item.placeID,
                        rawTypes: item.types,
                        distanceMeters: item.distanceMeters?.doubleValue
                    ))
                }
                print("4")
                completion(.success(result))
            }
        }
    }

    func getDetailById(placeId: String, fields: [PlaceFields], completion: @escaping (Result<PlaceItem?, Error>) -> Void) {
        var fieldValue = UInt(0)
        for item in fields {
            if let next = GmsPlacesApiImpl.fieldDict[item] {
                fieldValue = fieldValue | UInt(next.rawValue)
            }
        }
        let field = GMSPlaceField(rawValue: fieldValue)
        GMSPlacesClient.shared().fetchPlace(fromPlaceID: placeId, placeFields: field, sessionToken: GmsPlacesApiImpl.token) { place, error in
            if error != nil {
                completion(.failure(FlutterError(code: "error", message: error!.localizedDescription, details: error.debugDescription)))
            } else {
                if place == nil {
                    completion(.success(nil))
                } else {
                    var rawAddressComponents: [AddressComponent] = []
                    if place!.addressComponents != nil {
                        for item in place!.addressComponents! {
                            rawAddressComponents.append(AddressComponent(
                                name: item.name,
                                shortName: item.shortName,
                                rawTypes: item.types
                            ))
                        }
                    }
                    var plusCode: PlacePlusCode?
                    if let rawPlusCode = place!.plusCode {
                        plusCode = PlacePlusCode(
                            globalCode: rawPlusCode.globalCode,
                            compoundCode: rawPlusCode.compoundCode
                        )
                    }

                    var viewPort: PlaceViewport?
                    if let viewPortInfo = place!.viewportInfo {
                        viewPort = PlaceViewport(
                            northEast: PlaceCoordinate(
                                latitude: viewPortInfo.northEast.latitude,
                                longitude: viewPortInfo.northEast.longitude
                            ),
                            southWest: PlaceCoordinate(
                                latitude: viewPortInfo.southWest.latitude,
                                longitude: viewPortInfo.southWest.longitude
                            ),
                            valid: viewPortInfo.isValid
                        )
                    }

                    completion(.success(PlaceItem(
                        formattedAddress: place!.formattedAddress,
                        rawAddressComponents: rawAddressComponents,
                        businessStatus: GmsPlacesApiImpl.businessStatusDict[place!.businessStatus] ?? .unknown,
                        placeId: place!.placeID,
                        coordinate: PlaceCoordinate(
                            latitude: place!.coordinate.latitude,
                            longitude: place!.coordinate.longitude
                        ),
                        name: place!.name,
                        plusCode: plusCode,
                        rawTypes: place!.types,
                        viewport: viewPort
                    )))
                }
            }
        }
    }
}

class TokenEmptyError: Error {}

File: ex2/android/src/main/kotlin/dev/laihz/gms_places_api/GmsPlacesApiImpl.kt
package dev.laihz.gms_places_api

import android.content.Context
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.PlaceTypes
import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest

class GmsPlacesApiImpl(private val context: Context) : GmsPlacesApi {
    companion object {
        var token: AutocompleteSessionToken? = null
    }

    override fun ensureInitialized() {
        token = AutocompleteSessionToken.newInstance()
    }

    override fun autocomplete(
        fromQuery: String,
        filter: PlacesFilterType,
        callback: (Result<List<Prediction>>) -> Unit
    ) {
        val filter = when (filter) {
            PlacesFilterType.GEOCODE -> PlaceTypes.GEOCODE
            PlacesFilterType.ADDRESS -> PlaceTypes.ADDRESS
            PlacesFilterType.ESTABLISHMENT -> PlaceTypes.ESTABLISHMENT
            PlacesFilterType.REGION -> PlaceTypes.REGIONS
            PlacesFilterType.CITY -> PlaceTypes.ADDRESS
        }
        val req = FindAutocompletePredictionsRequest.builder()
            .setTypesFilter(listOf(filter))
            .setSessionToken(token)
            .setQuery(fromQuery)
            .build()

        Places.createClient(context).findAutocompletePredictions(req)
            .addOnSuccessListener { response ->
                val items = mutableListOf<Prediction>()
                for (item in response.autocompletePredictions) {
                    items.add(
                        Prediction(
                            attributed = PredictionAttributed(
                                fullText = item.getFullText(null).toString(),
                                primaryText = item.getPrimaryText(null).toString(),
                                secondaryText = item.getSecondaryText(null).toString(),
                            ),
                            placeID = item.placeId,
                            rawTypes = item.placeTypes.map { it -> it.name }
                        )
                    )
                }
                callback(Result.success(items))
            }
            .addOnFailureListener { exception ->
                callback(Result.failure(exception))
            }
    }

    override fun getDetailById(
        placeId: String,
        fields: List<PlaceFields>,
        callback: (Result<PlaceItem?>) -> Unit
    ) {
        val placeFields = fields.map { it ->
            when (it) {
                PlaceFields.FORMATTEDADDRESS -> Place.Field.ADDRESS
                PlaceFields.ADDRESSCOMPONENTS -> Place.Field.ADDRESS_COMPONENTS
                PlaceFields.BUSINESSSTATUS -> Place.Field.BUSINESS_STATUS
                PlaceFields.PLACEID -> Place.Field.ID
                PlaceFields.COORDINATE -> Place.Field.LAT_LNG
                PlaceFields.NAME -> Place.Field.NAME
                PlaceFields.PHOTOS -> Place.Field.PHOTO_METADATAS
                PlaceFields.PLUSCODE -> Place.Field.PLUS_CODE
                PlaceFields.TYPES -> Place.Field.TYPES
                PlaceFields.VIEWPORT -> Place.Field.VIEWPORT
            }
        }
        val req = FetchPlaceRequest.newInstance(
            placeId,
            placeFields
        )
        Places.createClient(context).fetchPlace(req)
            .addOnSuccessListener { response ->
                val place = response.place
                callback(
                    Result.success(
                        PlaceItem(
                            formattedAddress = place.address,
                            rawAddressComponents = place.addressComponents?.asList()?.map { it ->
                                AddressComponent(
                                    name = it.name,
                                    shortName = it.shortName,
                                    rawTypes = it.types
                                )
                            } ?: listOf(),
                            businessStatus = when(place.businessStatus) {
                                Place.BusinessStatus.OPERATIONAL -> PlacesBusinessStatus.OPERATIONAL
                                Place.BusinessStatus.CLOSED_TEMPORARILY -> PlacesBusinessStatus.CLOSEDTEMPORARILY
                                Place.BusinessStatus.CLOSED_PERMANENTLY -> PlacesBusinessStatus.CLOSEDTEMPORARILY
                                else -> PlacesBusinessStatus.OPERATIONAL
                            },
                            placeId = place.id,
                            coordinate = place.latLng?.let {
                                PlaceCoordinate(
                                    latitude = it.latitude,
                                    longitude = it.longitude
                                )
                            },
                            name = place.name,
                            plusCode = place.plusCode?.let {
                                PlacePlusCode(
                                    globalCode = it.globalCode ?: "",
                                    compoundCode = it.compoundCode
                                )
                            },
                            rawTypes = place.types?.map { it -> it.name }?.toList(),
                            viewport = place.viewport?.let {
                                PlaceViewport(
                                    northEast = PlaceCoordinate(
                                        latitude = it.northeast.latitude,
                                        longitude = it.northeast.longitude,
                                    ),
                                    southWest = PlaceCoordinate(
                                        latitude = it.southwest.latitude,
                                        longitude = it.southwest.longitude,
                                    ),
                                    valid = true,
                                )
                            }
                        )
                    )
                )
            }
            .addOnFailureListener { exception ->
                callback(Result.failure(exception))
            }
    }

}

